File: go/configa/k8s.go
package configa

import (
	"context"
	"crypto/tls"
	"crypto/x509"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"os"
	"strings"
	"sync"
	"time"
)

var (
	ErrSecretNotFound    = errors.New("configa: kubernetes secret not found")
	ErrSecretForbidden   = errors.New("configa: forbidden: insufficient permissions to read kubernetes secret")
	ErrSecretFetchFailed = errors.New("configa: failed to fetch kubernetes secret")
)

const (
	defaultSATokenPath     = "/var/run/secrets/kubernetes.io/serviceaccount/token"
	defaultSACAPath        = "/var/run/secrets/kubernetes.io/serviceaccount/ca.crt"
	defaultSANamespacePath = "/var/run/secrets/kubernetes.io/serviceaccount/namespace"
)

// InClusterSecretClient fetches Kubernetes secrets using the pod's in-cluster
// service account credentials. Credentials are loaded lazily on the first
// GetSecret call so the client is safe to construct outside Kubernetes.
type InClusterSecretClient struct {
	once      sync.Once
	initErr   error
	client    *http.Client
	token     string
	apiBase   string
	defaultNS string

	// Overridable paths — set by NewInClusterSecretClient; tests may override
	// by constructing the struct directly (same package).
	tokenPath     string
	caPath        string
	namespacePath string
}

// NewInClusterSecretClient returns a SecretClient backed by the Kubernetes
// in-cluster service account. Construction never fails — any errors in reading
// service account files surface on the first GetSecret call.
func NewInClusterSecretClient() SecretClient {
	return &InClusterSecretClient{
		tokenPath:     defaultSATokenPath,
		caPath:        defaultSACAPath,
		namespacePath: defaultSANamespacePath,
	}
}

// k8sSecretResponse is the subset of the Kubernetes v1.Secret JSON we need.
// encoding/json automatically base64-decodes []byte fields from JSON strings.
type k8sSecretResponse struct {
	Data map[string][]byte `json:"data"`
}

func (c *InClusterSecretClient) loadCredentials() {
	c.once.Do(func() {
		tokenBytes, err := os.ReadFile(c.tokenPath)
		if err != nil {
			c.initErr = fmt.Errorf("configa: read service account token: %w", err)
			return
		}

		caBytes, err := os.ReadFile(c.caPath)
		if err != nil {
			c.initErr = fmt.Errorf("configa: read service account CA cert: %w", err)
			return
		}

		pool := x509.NewCertPool()
		if !pool.AppendCertsFromPEM(caBytes) {
			c.initErr = fmt.Errorf("configa: failed to parse service account CA cert")
			return
		}

		nsBytes, err := os.ReadFile(c.namespacePath)
		if err != nil {
			c.initErr = fmt.Errorf("configa: read service account namespace: %w", err)
			return
		}

		host := os.Getenv("KUBERNETES_SERVICE_HOST")
		port := os.Getenv("KUBERNETES_SERVICE_PORT")
		if host == "" || port == "" {
			c.initErr = fmt.Errorf("configa: KUBERNETES_SERVICE_HOST or KUBERNETES_SERVICE_PORT not set")
			return
		}

		c.token = strings.TrimSpace(string(tokenBytes))
		c.defaultNS = strings.TrimSpace(string(nsBytes))
		c.apiBase = fmt.Sprintf("https://%s:%s", host, port)
		c.client = &http.Client{
			Timeout: 10 * time.Second,
			Transport: &http.Transport{
				TLSClientConfig: &tls.Config{
					RootCAs: pool,
				},
			},
		}
	})
}

// GetSecret fetches all data keys for the named secret.
// When namespace is empty the service account's own namespace is used.
func (c *InClusterSecretClient) GetSecret(ctx context.Context, namespace, name string) (map[string]string, error) {
	c.loadCredentials()
	if c.initErr != nil {
		return nil, c.initErr
	}

	ns := namespace
	if ns == "" {
		ns = c.defaultNS
	}

	url := fmt.Sprintf("%s/api/v1/namespaces/%s/secrets/%s", c.apiBase, ns, name)
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return nil, fmt.Errorf("configa: build k8s request: %w", err)
	}
	req.Header.Set("Authorization", "Bearer "+c.token)

	resp, err := c.client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("configa: k8s request failed: %w", err)
	}
	defer resp.Body.Close()

	switch resp.StatusCode {
	case http.StatusOK:
		// handled below
	case http.StatusNotFound:
		return nil, fmt.Errorf("%w: %s/%s", ErrSecretNotFound, ns, name)
	case http.StatusForbidden, http.StatusUnauthorized:
		return nil, fmt.Errorf("%w: %s/%s", ErrSecretForbidden, ns, name)
	default:
		return nil, fmt.Errorf("%w: %s/%s (HTTP %d)", ErrSecretFetchFailed, ns, name, resp.StatusCode)
	}

	var secret k8sSecretResponse
	if err := json.NewDecoder(resp.Body).Decode(&secret); err != nil {
		return nil, fmt.Errorf("configa: decode k8s secret response: %w", err)
	}

	result := make(map[string]string, len(secret.Data))
	for k, v := range secret.Data {
		result[k] = string(v)
	}

	return result, nil
}

File: typescript/packages/configa/src/k8s.ts
import { readFile } from 'node:fs/promises'
import { Agent, fetch as undiciFetch } from 'undici'
import {
  ErrSecretFetchFailed,
  ErrSecretForbidden,
  ErrSecretNotFound
} from './errors.js'
import type { SecretClient } from './types.js'

const DEFAULT_TOKEN_PATH =
  '/var/run/secrets/kubernetes.io/serviceaccount/token'
const DEFAULT_CA_PATH = '/var/run/secrets/kubernetes.io/serviceaccount/ca.crt'
const DEFAULT_NAMESPACE_PATH =
  '/var/run/secrets/kubernetes.io/serviceaccount/namespace'

const REQUEST_TIMEOUT_MS = 10_000

export interface InClusterSecretClientPaths {
  tokenPath?: string
  caPath?: string
  namespacePath?: string
}

interface Credentials {
  token: string
  defaultNamespace: string
  apiBase: string
  dispatcher: Agent
}

interface K8sSecretResponse {
  data?: Record<string, string>
}

/**
 * Fetches Kubernetes secrets using the pod's in-cluster service account
 * credentials. Credentials are loaded lazily on the first getSecret() call,
 * so the client is safe to construct outside Kubernetes. Port of
 * go/configa/k8s.go's InClusterSecretClient.
 */
export class InClusterSecretClient implements SecretClient {
  private readonly tokenPath: string
  private readonly caPath: string
  private readonly namespacePath: string
  private credentialsPromise: Promise<Credentials> | undefined

  constructor(paths: InClusterSecretClientPaths = {}) {
    this.tokenPath = paths.tokenPath ?? DEFAULT_TOKEN_PATH
    this.caPath = paths.caPath ?? DEFAULT_CA_PATH
    this.namespacePath = paths.namespacePath ?? DEFAULT_NAMESPACE_PATH
  }

  private loadCredentials(): Promise<Credentials> {
    // Memoized promise plays the role of Go's sync.Once: credential files
    // are read at most once, even under concurrent getSecret() calls.
    if (!this.credentialsPromise) {
      this.credentialsPromise = this.doLoadCredentials()
    }
    return this.credentialsPromise
  }

  private async doLoadCredentials(): Promise<Credentials> {
    let token: string
    try {
      token = (await readFile(this.tokenPath, 'utf8')).trim()
    } catch (err) {
      throw new Error(
        `configa: read service account token: ${(err as Error).message}`,
        { cause: err }
      )
    }

    let ca: string
    try {
      ca = await readFile(this.caPath, 'utf8')
    } catch (err) {
      throw new Error(
        `configa: read service account CA cert: ${(err as Error).message}`,
        { cause: err }
      )
    }

    let defaultNamespace: string
    try {
      defaultNamespace = (await readFile(this.namespacePath, 'utf8')).trim()
    } catch (err) {
      throw new Error(
        `configa: read service account namespace: ${(err as Error).message}`,
        { cause: err }
      )
    }

    const host = process.env.KUBERNETES_SERVICE_HOST
    const port = process.env.KUBERNETES_SERVICE_PORT
    if (!host || !port) {
      throw new Error(
        'configa: KUBERNETES_SERVICE_HOST or KUBERNETES_SERVICE_PORT not set'
      )
    }

    return {
      token,
      defaultNamespace,
      apiBase: `https://${host}:${port}`,
      dispatcher: new Agent({ connect: { ca } })
    }
  }

  async getSecret(
    namespace: string,
    name: string
  ): Promise<Record<string, string>> {
    const creds = await this.loadCredentials()
    const ns = namespace || creds.defaultNamespace

    const url = `${creds.apiBase}/api/v1/namespaces/${ns}/secrets/${name}`
    let response: Awaited<ReturnType<typeof undiciFetch>>
    try {
      response = await undiciFetch(url, {
        headers: { authorization: `Bearer ${creds.token}` },
        dispatcher: creds.dispatcher,
        signal: AbortSignal.timeout(REQUEST_TIMEOUT_MS)
      })
    } catch (err) {
      throw new Error(
        `configa: k8s request failed: ${(err as Error).message}`,
        { cause: err }
      )
    }

    if (response.status === 404) {
      throw new ErrSecretNotFound(ns, name)
    }
    if (response.status === 403 || response.status === 401) {
      throw new ErrSecretForbidden(ns, name)
    }
    if (response.status !== 200) {
      throw new ErrSecretFetchFailed(ns, name, `HTTP ${response.status}`)
    }

    const body = (await response.json()) as K8sSecretResponse
    const result: Record<string, string> = {}
    for (const [key, value] of Object.entries(body.data ?? {})) {
      result[key] = Buffer.from(value, 'base64').toString('utf8')
    }
    return result
  }
}

/** Returns a SecretClient backed by the Kubernetes in-cluster service account. */
export function newInClusterSecretClient(): SecretClient {
  return new InClusterSecretClient()
}
